tools/codegen/codespec/attribute.go

```diff
@@ -160,6 +160,48 @@ func (s *APISpecSchema) buildBoolAttr(name string, computability ComputedOptiona
 	return result, nil
 }
 
+// buildRegularCollectionAttribute creates an attribute for list/set of regular (non-nested) element types.
+func (s *APISpecSchema) buildRegularCollectionAttribute(name, tfModelName string, computability ComputedOptionalRequired, isSet bool, elemType ElemType) *Attribute {
+	attr := &Attribute{
+		TFSchemaName:             stringcase.ToSnakeCase(name),
+		TFModelName:              tfModelName,
+		ComputedOptionalRequired: computability,
+		DeprecationMessage:       s.GetDeprecationMessage(),
+		Description:              s.GetDescription(),
+	}
+
+	if isSet {
+		attr.CustomType = NewCustomSetType(elemType)
+		attr.Set = &SetAttribute{ElementType: elemType}
+	} else {
+		attr.CustomType = NewCustomListType(elemType)
+		attr.List = &ListAttribute{ElementType: elemType}
+	}
+
+	return attr
+}
+
+// buildNestedCollectionAttribute creates an attribute for list/set of nested objects (with attributes).
+func (s *APISpecSchema) buildNestedCollectionAttribute(name, tfModelName, nestedObjectName string, computability ComputedOptionalRequired, isSet bool, nestedObject NestedAttributeObject) *Attribute {
+	attr := &Attribute{
+		TFSchemaName:             stringcase.ToSnakeCase(name),
+		TFModelName:              tfModelName,
+		ComputedOptionalRequired: computability,
+		DeprecationMessage:       s.GetDeprecationMessage(),
+		Description:              s.GetDescription(),
+	}
+
+	if isSet {
+		attr.CustomType = NewCustomNestedSetType(nestedObjectName)
+		attr.SetNested = &SetNestedAttribute{NestedObject: nestedObject}
+	} else {
+		attr.CustomType = NewCustomNestedListType(nestedObjectName)
+		attr.ListNested = &ListNestedAttribute{NestedObject: nestedObject}
+	}
```

Comment:
```suggestion
	attr.CustomType = NewCustomNestedListType(nestedObjectName)
	attr.ListNested = &ListNestedAttribute{NestedObject: nestedObject}
	if isSet {
		attr.CustomType = NewCustomNestedSetType(nestedObjectName)
		attr.SetNested = &SetNestedAttribute{NestedObject: nestedObject}
	} 
```

nit: don't know if it have side effect to update the values and then set again if set

Reply:
I would go for the current implementation which does not allocate more than necessary (even if the compiler does optimize this, which I don't think it does).